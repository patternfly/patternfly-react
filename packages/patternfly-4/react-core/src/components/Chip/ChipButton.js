import React from 'react';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import styles from '@patternfly/patternfly/components/Chip/chip.css';
import { Button } from '../Button';

const ChipButton = ({ ariaLabel, children, className, onClick, ...props }) => {
  return (
    <Button variant="plain" aria-label={ariaLabel} onClick={onClick} className={className} {...props}>
      {children}
    </Button>
  );
};

ChipButton.propTypes = {
  /** Aria label for chip button */
  ariaLabel: PropTypes.string,
  /** Content rendered inside the chip item */
  children: PropTypes.node,
  /** Additional classes added to the chip item */
  className: PropTypes.string,
  /** Function that is called when clicking on the chip button */
  onClick: PropTypes.func,
};

ChipButton.defaultProps = {
  ariaLabel: 'close',
  children: null,
  className: '',
  onClick: () => { }
};

export default ChipButton;
