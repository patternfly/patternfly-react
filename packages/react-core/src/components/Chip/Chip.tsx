import * as React from 'react';
import { css } from '@patternfly/react-styles';
import { Button } from '../Button';
import { Tooltip, TooltipPosition } from '../Tooltip';
import TimesIcon from '@patternfly/react-icons/dist/esm/icons/times-icon';
import styles from '@patternfly/react-styles/css/components/Chip/chip';
import { GenerateId } from '../../helpers/GenerateId/GenerateId';
import { getOUIAProps, OUIAProps, getDefaultOUIAId } from '../../helpers';

export interface ChipProps extends React.HTMLProps<HTMLDivElement>, OUIAProps {
  /** Badge to add to the chip. The badge will be rendered after the chip text. */
  badge?: React.ReactNode;
  /** Content rendered inside the chip text */
  children?: React.ReactNode;
  /** Aria Label for close button */
  closeBtnAriaLabel?: string;
  /** Additional classes added to the chip item */
  className?: string;
  /** Flag indicating if the chip is an overflow chip */
  isOverflowChip?: boolean;
  /** Flag indicating if chip is read only */
  isReadOnly?: boolean;
  /** Function that is called when clicking on the chip close button */
  onClick?: (event: React.MouseEvent) => void;
  /** Component that will be used for chip. It is recommended that <button> or <li>  are used when the chip is an overflow chip. */
  component?: React.ReactNode;
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition?:
    | TooltipPosition
    | 'auto'
    | 'top'
    | 'bottom'
    | 'left'
    | 'right'
    | 'top-start'
    | 'top-end'
    | 'bottom-start'
    | 'bottom-end'
    | 'left-start'
    | 'left-end'
    | 'right-start'
    | 'right-end';

  /** Css property expressed in percentage or any css unit that overrides the default value of the max-width of the chip's text */
  textMaxWidth?: string;
  /** Value to overwrite the randomly generated data-ouia-component-id.*/
  ouiaId?: number | string;
}

interface ChipState {
  isTooltipVisible: boolean;
  ouiaStateId: string;
}

class Chip extends React.Component<ChipProps, ChipState> {
  static displayName = 'Chip';
  constructor(props: ChipProps) {
    super(props);
    this.state = {
      isTooltipVisible: false,
      ouiaStateId: getDefaultOUIAId(Chip.displayName)
    };
  }
  span = React.createRef<HTMLSpanElement>();

  static defaultProps: ChipProps = {
    closeBtnAriaLabel: 'close',
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top' as 'auto' | 'top' | 'bottom' | 'left' | 'right',
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onClick: (_e: React.MouseEvent) => undefined as any,
    component: 'div' as React.ReactNode
  };

  componentDidMount() {
    this.setState({
      isTooltipVisible: Boolean(this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth)
    });
  }

  setChipStyle = () => ({
    '--pf-v5-c-chip__text--MaxWidth': this.props.textMaxWidth
  });

  renderOverflowChip = () => {
    const {
      badge,
      children,
      className,
      onClick,
      ouiaId,
      textMaxWidth,
      style,
      component,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      tooltipPosition,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      isOverflowChip,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      closeBtnAriaLabel,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      isReadOnly,
      ...props
    } = this.props;
    const Component = component as any;
    return (
      <Component
        onClick={onClick}
        {...(textMaxWidth && {
          style: this.setChipStyle(),
          ...style
        })}
        className={css(styles.chip, styles.modifiers.overflow, className)}
        {...(component === 'button' ? { type: 'button' } : {})}
        {...getOUIAProps('OverflowChip', ouiaId !== undefined ? ouiaId : this.state.ouiaStateId)}
        {...props}
      >
        <span className={css(styles.chipContent)}>
          <span className={css(styles.chipText)}>{children}</span>
          {badge && badge}
        </span>
      </Component>
    );
  };

  renderInnerChip(id: string) {
    const {
      badge,
      children,
      className,
      onClick,
      closeBtnAriaLabel,
      isReadOnly,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      isOverflowChip,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      tooltipPosition,
      component,
      ouiaId,
      ...props
    } = this.props;
    const Component = component as any;
    return (
      <Component
        {...(this.props.textMaxWidth && {
          style: this.setChipStyle()
        })}
        className={css(styles.chip, className)}
        {...(this.state.isTooltipVisible && { tabIndex: 0 })}
        {...getOUIAProps(Chip.displayName, ouiaId !== undefined ? ouiaId : this.state.ouiaStateId)}
        {...props}
      >
        <span className={css(styles.chipContent)}>
          <span ref={this.span} className={css(styles.chipText)} id={id}>
            {children}
          </span>
          {badge && badge}
        </span>
        {!isReadOnly && (
          <span className={css(styles.chipActions)}>
            <Button
              onClick={onClick}
              variant="plain"
              aria-label={closeBtnAriaLabel}
              id={`remove_${id}`}
              aria-labelledby={`remove_${id} ${id}`}
              ouiaId={ouiaId || closeBtnAriaLabel}
            >
              <TimesIcon aria-hidden="true" />
            </Button>
          </span>
        )}
      </Component>
    );
  }

  renderChip = (randomId: string) => {
    const { children, tooltipPosition } = this.props;
    if (this.state.isTooltipVisible) {
      return (
        <Tooltip position={tooltipPosition} content={children}>
          {this.renderInnerChip(randomId)}
        </Tooltip>
      );
    }
    return this.renderInnerChip(randomId);
  };

  render() {
    const { isOverflowChip } = this.props;
    return (
      <GenerateId>
        {(randomId) => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(this.props.id || randomId))}
      </GenerateId>
    );
  }
}

export { Chip };
