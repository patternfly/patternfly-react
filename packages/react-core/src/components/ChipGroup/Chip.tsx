import * as React from 'react';
import { css } from '@patternfly/react-styles';
import { Button } from '../Button';
import { Tooltip } from '../Tooltip';
import TimesIcon from '@patternfly/react-icons/dist/js/icons/times-icon';
import styles from '@patternfly/react-styles/css/components/Chip/chip';
import { GenerateId } from '../../helpers/GenerateId/GenerateId';
import { getOUIAProps, OUIAProps, getDefaultOUIAId } from '../../helpers';

export interface ChipProps extends React.HTMLProps<HTMLDivElement>, OUIAProps {
  /** Content rendered inside the chip text */
  children?: React.ReactNode;
  /** Aria Label for close button */
  closeBtnAriaLabel?: string;
  /** Additional classes added to the chip item */
  className?: string;
  /** Flag indicating if the chip is an overflow chip */
  isOverflowChip?: boolean;
  /** Flag indicating if chip is read only */
  isReadOnly?: boolean;
  /** Function that is called when clicking on the chip close button */
  onClick?: (event: React.MouseEvent) => void;
  /** Component that will be used for chip. It is recommended that <button> or <li>  are used when the chip is an overflow chip. */
  component?: React.ReactNode;
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition?: 'auto' | 'top' | 'bottom' | 'left' | 'right';
}

interface ChipState {
  isTooltipVisible: boolean;
  ouiaStateId: string;
}

export class Chip extends React.Component<ChipProps, ChipState> {
  static displayName = 'Chip';
  constructor(props: ChipProps) {
    super(props);
    this.state = {
      isTooltipVisible: false,
      ouiaStateId: getDefaultOUIAId(Chip.displayName)
    };
  }
  span = React.createRef<HTMLSpanElement>();

  static defaultProps: ChipProps = {
    closeBtnAriaLabel: 'close',
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top' as 'auto' | 'top' | 'bottom' | 'left' | 'right',
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onClick: (_e: React.MouseEvent) => undefined as any,
    component: 'div' as React.ReactNode
  };

  componentDidMount() {
    this.setState({
      isTooltipVisible: Boolean(this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth)
    });
  }

  renderOverflowChip = () => {
    const { children, className, onClick, ouiaId } = this.props;
    const Component = this.props.component as any;
    return (
      <Component
        onClick={onClick}
        className={css(styles.chip, styles.modifiers.overflow, className)}
        {...(this.props.component === 'button' ? { type: 'button' } : {})}
        {...getOUIAProps('OverflowChip', ouiaId !== undefined ? ouiaId : this.state.ouiaStateId)}
      >
        <span className={css(styles.chipText)}>{children}</span>
      </Component>
    );
  };

  renderInnerChip(id: string) {
    const { children, className, onClick, closeBtnAriaLabel, isReadOnly, component, ouiaId } = this.props;
    const Component = component as any;
    return (
      <Component
        className={css(styles.chip, className)}
        {...(this.state.isTooltipVisible && { tabIndex: 0 })}
        {...getOUIAProps(Chip.displayName, ouiaId !== undefined ? ouiaId : this.state.ouiaStateId)}
      >
        <span ref={this.span} className={css(styles.chipText)} id={id}>
          {children}
        </span>
        {!isReadOnly && (
          <Button
            onClick={onClick}
            variant="plain"
            aria-label={closeBtnAriaLabel}
            id={`remove_${id}`}
            aria-labelledby={`remove_${id} ${id}`}
            ouiaId={ouiaId || closeBtnAriaLabel}
          >
            <TimesIcon aria-hidden="true" />
          </Button>
        )}
      </Component>
    );
  }

  renderChip = (randomId: string) => {
    const { children, tooltipPosition } = this.props;
    if (this.state.isTooltipVisible) {
      return (
        <Tooltip position={tooltipPosition} content={children}>
          {this.renderInnerChip(randomId)}
        </Tooltip>
      );
    }
    return this.renderInnerChip(randomId);
  };

  render() {
    const { isOverflowChip } = this.props;
    return (
      <GenerateId>
        {randomId => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(this.props.id || randomId))}
      </GenerateId>
    );
  }
}
