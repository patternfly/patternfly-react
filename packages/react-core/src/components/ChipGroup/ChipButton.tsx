import * as React from 'react';
import { Button, ButtonProps } from '../Button';

export interface ChipButtonProps extends ButtonProps {
  /** Aria label for chip button */
  ariaLabel?: string;
  /** Content rendered inside the chip item */
  children?: React.ReactNode;
  /** Additional classes added to the chip item */
  className?: string;
  /** Function that is called when clicking on the chip button */
  onClick?: (event: React.MouseEvent) => void;
}

export const ChipButton: React.FunctionComponent<ChipButtonProps> = ({
  ariaLabel = 'close',
  children = null,
  className = '',
  onClick = () => undefined,
  ...props
}: ChipButtonProps) => (
  <Button variant="plain" aria-label={ariaLabel} onClick={onClick} className={className} {...props}>
    {children}
  </Button>
);
