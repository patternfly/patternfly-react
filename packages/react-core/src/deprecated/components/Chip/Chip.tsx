import * as React from 'react';
import { TooltipPosition } from '../../../components/Tooltip';
import { Label, LabelProps } from '../../../components/Label';
import { getOUIAProps, OUIAProps, getDefaultOUIAId } from '../../../helpers';

export interface ChipProps extends LabelProps, OUIAProps {
  /** Badge to add to the chip. The badge will be rendered after the chip text. */
  badge?: React.ReactNode;
  /** Content rendered inside the chip text */
  children?: React.ReactNode;
  /** Aria Label for close button */
  closeBtnAriaLabel?: string;
  /** Additional classes added to the chip item */
  className?: string;
  /** Flag indicating if the chip is an overflow chip */
  isOverflowChip?: boolean;
  /** Flag indicating if chip is read only */
  isReadOnly?: boolean;
  /** Function that is called when clicking on the chip close button */
  onClick?: (event: React.MouseEvent) => void;
  /** @deprecated Component that will be used for chip. It is recommended that <button> or <li>  are used when the chip is an overflow chip. */
  component?: React.ReactNode;
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition?:
    | TooltipPosition
    | 'auto'
    | 'top'
    | 'bottom'
    | 'left'
    | 'right'
    | 'top-start'
    | 'top-end'
    | 'bottom-start'
    | 'bottom-end'
    | 'left-start'
    | 'left-end'
    | 'right-start'
    | 'right-end';

  /** Css property expressed in percentage or any css unit that overrides the default value of the max-width of the chip's text */
  textMaxWidth?: string;
  /** Value to overwrite the randomly generated data-ouia-component-id.*/
  ouiaId?: number | string;
}

class Chip extends React.Component<ChipProps> {
  static displayName = 'Chip';
  constructor(props: ChipProps) {
    super(props);
  }

  static defaultProps: ChipProps = {
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top' as 'auto' | 'top' | 'bottom' | 'left' | 'right',

    onClick: (_e: React.MouseEvent) => undefined as any,
    textMaxWidth: '16ch'
  };

  render() {
    const {
      isOverflowChip,
      children,
      badge,
      onClick,
      className,
      textMaxWidth,
      isReadOnly,
      closeBtnAriaLabel,
      tooltipPosition,
      ouiaId,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      component,
      ...props
    } = this.props;

    return (
      <Label
        variant="outline"
        className={className}
        textMaxWidth={textMaxWidth}
        tooltipPosition={tooltipPosition}
        {...(!isReadOnly &&
          !isOverflowChip && {
            onClose: onClick,
            closeBtnAriaLabel: closeBtnAriaLabel ? closeBtnAriaLabel : `Close ${children}`
          })}
        {...(isOverflowChip && { onClick, variant: 'overflow' })}
        {...getOUIAProps(
          isOverflowChip ? 'OverflowChip' : Chip.displayName,
          ouiaId !== undefined ? ouiaId : getDefaultOUIAId(Chip.displayName)
        )}
        {...props}
      >
        {children}
        {badge && <React.Fragment> {badge}</React.Fragment>}
      </Label>
    );
  }
}

export { Chip };
