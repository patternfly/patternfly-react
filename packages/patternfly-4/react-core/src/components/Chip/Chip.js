import React from 'react';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import ChipButton from './ChipButton';
import { Tooltip, TooltipPosition } from '../Tooltip';
import { TimesCircleIcon } from '@patternfly/react-icons';
import styles from '@patternfly/patternfly-next/components/Chip/chip.css';
import GenerateId from '../../internal/GenerateId/GenerateId';
class Chip extends React.Component {
  span = React.createRef();
  state = { isTooltipVisible: false };

  componentDidMount() {
    this.setState({ isTooltipVisible: this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth });
  }

  renderOverflowChip = () => {
    const { children, className, onClick } = this.props;
    return (
      <div className={css(styles.chip, styles.modifiers.overflow, className)}>
        <ChipButton onClick={onClick}>
          <span className={css(styles.chipText)}>{children}</span>
        </ChipButton>
      </div>
    );
  }

  renderChip = (randomId) => {
    const {
      children,
      closeBtnAriaLabel,
      tooltipPosition,
      className,
      onClick,
    } = this.props;
    if (this.state.isTooltipVisible) {
      return (
        <Tooltip position={tooltipPosition} content={children}>
          <div className={css(styles.chip, className)}>
            <span ref={this.span} className={css(styles.chipText)} id={randomId}>
              {children}
            </span>
            <ChipButton onClick={onClick} ariaLabel={closeBtnAriaLabel} id={`remove_${randomId}`} aria-labelledby={`remove_${randomId} ${randomId}`}>
              <TimesCircleIcon aria-hidden="true" />
            </ChipButton>
          </div>
        </Tooltip>
      )
    } else {
      return (
        <div className={css(styles.chip, className)}>
          <span ref={this.span} className={css(styles.chipText)} id={randomId}>
            {children}
          </span>
          <ChipButton onClick={onClick} ariaLabel={closeBtnAriaLabel} id={`remove_${randomId}`} aria-labelledby={`remove_${randomId} ${randomId}`}>
            <TimesCircleIcon aria-hidden="true" />
          </ChipButton>
        </div>
      )
    }
  }

  render() {
    const {
      isOverflowChip,
    } = this.props;
    return (
      <GenerateId>
        {(randomId) =>
          (
            <React.Fragment>
              {isOverflowChip ? this.renderOverflowChip() : this.renderChip(randomId)}
            </React.Fragment>
          )
        }
      </GenerateId>
    )
  }
}
Chip.propTypes = {
  /** Content rendered inside the chip text */
  children: PropTypes.string,
  /** Aria Label for close button */
  closeBtnAriaLabel: PropTypes.string,
  /** ID of the chip */
  id: PropTypes.string,
  /** Additional classes added to the chip item */
  className: PropTypes.string,
  /** Flag indicating if the chip has overflow*/
  isOverflowChip: PropTypes.bool,
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition: PropTypes.oneOf(Object.values(TooltipPosition)),
  /** Function that is called when clicking on the chip button */
  onClick: PropTypes.func,
};

Chip.defaultProps = {
  id: undefined,
  closeBtnAriaLabel: 'close',
  className: '',
  tooltipPosition: 'top',
  isOverflowChip: false,
};

export default Chip;
