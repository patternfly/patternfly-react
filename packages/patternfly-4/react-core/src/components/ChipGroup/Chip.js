import React from 'react';
import { css } from '@patternfly/react-styles';
import PropTypes from 'prop-types';
import ChipButton from './ChipButton';
import { Tooltip, TooltipPosition } from '../Tooltip';
import { TimesCircleIcon } from '@patternfly/react-icons';
import styles from '@patternfly/patternfly/components/Chip/chip.css';
import GenerateId from '../../helpers/GenerateId/GenerateId';

class Chip extends React.Component {
  span = React.createRef();
  state = { isTooltipVisible: false };

  componentDidMount() {
    this.setState({
      isTooltipVisible: this.span.current && this.span.current.offsetWidth < this.span.current.scrollWidth
    });
  }

  renderOverflowChip = () => {
    const { children, className, onClick } = this.props;
    return (
      <div className={css(styles.chip, styles.modifiers.overflow, className)}>
        <ChipButton onClick={onClick}>
          <span className={css(styles.chipText)}>{children}</span>
        </ChipButton>
      </div>
    );
  };

  renderChip = randomId => {
    const { children, closeBtnAriaLabel, tooltipPosition, className, onClick } = this.props;
    if (this.state.isTooltipVisible) {
      return (
        <Tooltip position={tooltipPosition} content={children}>
          <li className={css(styles.chip, className)}>
            <span ref={this.span} className={css(styles.chipText)} id={randomId}>
              {children}
            </span>
            <ChipButton
              onClick={onClick}
              ariaLabel={closeBtnAriaLabel}
              id={`remove_${randomId}`}
              aria-labelledby={`remove_${randomId} ${randomId}`}
            >
              <TimesCircleIcon aria-hidden="true" />
            </ChipButton>
          </li>
        </Tooltip>
      );
    }
    return (
      <li className={css(styles.chip, className)}>
        <span ref={this.span} className={css(styles.chipText)} id={randomId}>
          {children}
        </span>
        <ChipButton
          onClick={onClick}
          ariaLabel={closeBtnAriaLabel}
          id={`remove_${randomId}`}
          aria-labelledby={`remove_${randomId} ${randomId}`}
        >
          <TimesCircleIcon aria-hidden="true" />
        </ChipButton>
      </li>
    );
  };

  render() {
    const { isOverflowChip } = this.props;
    return (
      <GenerateId>{randomId => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(randomId))}</GenerateId>
    );
  }
}
Chip.propTypes = {
  /** Content rendered inside the chip text */
  children: PropTypes.node,
  /** Aria Label for close button */
  closeBtnAriaLabel: PropTypes.string,
  /** Additional classes added to the chip item */
  className: PropTypes.string,
  /** Flag indicating if the chip has overflow */
  isOverflowChip: PropTypes.bool,
  /** Function that is called when clicking on the chip button */
  onClick: PropTypes.func,
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition: PropTypes.oneOf(Object.values(TooltipPosition))
};

Chip.defaultProps = {
  children: null,
  closeBtnAriaLabel: 'close',
  className: '',
  isOverflowChip: false,
  tooltipPosition: 'top'
};

export default Chip;
