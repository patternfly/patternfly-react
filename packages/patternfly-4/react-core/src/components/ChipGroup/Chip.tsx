import * as React from 'react';
import { css } from '@patternfly/react-styles';
import { ChipButton } from './ChipButton';
import { OneOf } from '../../helpers/typeUtils';
import { Tooltip, TooltipPosition } from '../Tooltip';
import { TimesCircleIcon } from '@patternfly/react-icons';
import styles from '@patternfly/react-styles/css/components/Chip/chip';
import GenerateId from '../../helpers/GenerateId/GenerateId';

export interface ChipProps extends React.HTMLProps<HTMLDivElement> {
  /** Content rendered inside the chip text */
  children?: React.ReactNode;
  /** Aria Label for close button */
  closeBtnAriaLabel?: string;
  /** Additional classes added to the chip item */
  className?: string; 
  /** Flag indicating if the chip has overflow */
  isOverflowChip?: boolean;
  /** Flag if chip is read only */
  isReadOnly?: boolean;
  /** Function that is called when clicking on the chip button */
  onClick?: (event: React.MouseEvent) => void
  /** Internal flag for which component will be used for chip */
  component?: React.ReactNode; 
  /** Position of the tooltip which is displayed if text is longer */
  tooltipPosition?: 'top' | 'bottom' | 'left' | 'right';
}

interface ChipState {
  isTooltipVisible: boolean;
}

export class Chip extends React.Component<ChipProps, ChipState> {
  constructor(props: ChipProps){
    super(props); 
    this.state = {
      isTooltipVisible: false
    }
  }
  span = React.createRef<HTMLSpanElement>(); 

  static defaultProps = {
    closeBtnAriaLabel: 'close',
    className: '',
    isOverflowChip: false,
    isReadOnly: false,
    tooltipPosition: 'top',
    onClick: () => undefined as any,
    component: 'div'
  }

  componentDidMount() {
    this.setState({
      isTooltipVisible: Boolean(this.span.current && 
        this.span.current.offsetWidth < this.span.current.scrollWidth)
    });
  }

  renderOverflowChip = () => {
    const { children, className, onClick } = this.props;
    const Component = this.props.component as any;
    return (
      <Component className={css(styles.chip, styles.modifiers.overflow, className)}>
        <ChipButton onClick={onClick}>
          <span className={css(styles.chipText)}>{children}</span>
        </ChipButton>
      </Component>
    );
  };

  renderChip = (randomId: string) => {
    const {
      children,
      closeBtnAriaLabel,
      tooltipPosition,
      className,
      onClick,
      isReadOnly
    } = this.props;
    const Component = this.props.component as any;
    if (this.state.isTooltipVisible) {
      return (
        <Tooltip position={tooltipPosition} content={children}>
          <Component className={css(styles.chip, isReadOnly && styles.modifiers.readOnly, className)}>
            <span ref={this.span} className={css(styles.chipText)} id={randomId}>
              {children}
            </span>
            {!isReadOnly && (
              <ChipButton
                onClick={onClick}
                ariaLabel={closeBtnAriaLabel}
                id={`remove_${randomId}`}
                aria-labelledby={`remove_${randomId} ${randomId}`}
              >
                <TimesCircleIcon aria-hidden="true" />
              </ChipButton>
            )}
          </Component>
        </Tooltip>
      );
    }
    return (
      <Component className={css(styles.chip, isReadOnly && styles.modifiers.readOnly, className)}>
        <span ref={this.span} className={css(styles.chipText)} id={randomId}>
          {children}
        </span>
        {!isReadOnly && (
          <ChipButton
            onClick={onClick}
            ariaLabel={closeBtnAriaLabel}
            id={`remove_${randomId}`}
            aria-labelledby={`remove_${randomId} ${randomId}`}
          >
            <TimesCircleIcon aria-hidden="true" />
          </ChipButton>
        )}
      </Component>
    );
  };
  
  render(){
    const { isOverflowChip } = this.props;
    return (
      <GenerateId>{randomId => (isOverflowChip ? this.renderOverflowChip() : this.renderChip(randomId))}</GenerateId>
    );
  }
};
