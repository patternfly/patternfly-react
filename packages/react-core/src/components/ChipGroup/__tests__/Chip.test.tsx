import React from 'react';
import { mount } from 'enzyme';
import { ChipButton } from '../ChipButton';
import { Chip } from '../Chip';

test('ChipButton', () => {
  const view = mount(
    <ChipButton id="my-chip-button" className="chip-bttn-cls">
      <b>Close</b>
    </ChipButton>
  );
  expect(view).toMatchSnapshot();
});

describe('Chip', () => {
  test('overflow', () => {
    const view = mount(
      <Chip className="my-chp-cls" isOverflowChip>
        4 more
      </Chip>
    );
    expect(view).toMatchSnapshot();
  });

  test('closable', () => {
    const view = mount(
      <Chip className="my-chp-cls" id="chip_one">
        Chip
      </Chip>
    );
    expect(view).toMatchSnapshot();
  });

  test('closable with tooltip', () => {
    const view = mount(
      <Chip className="my-chp-cls" id="chip_one">
        1234567890123456789
      </Chip>
    );
    expect(view).toMatchSnapshot();
  });

  test('readonly', () => {
    const view = mount(
      <Chip className="my-chp-cls" isReadOnly>
        4 more
      </Chip>
    );
    expect(view).toMatchSnapshot();
  });
});
