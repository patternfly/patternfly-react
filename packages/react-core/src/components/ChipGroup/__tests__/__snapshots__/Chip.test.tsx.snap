// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip closable 1`] = `
<Chip
  className="my-chp-cls"
  closeBtnAriaLabel="close"
  component="div"
  id="chip_one"
  isOverflowChip={false}
  isReadOnly={false}
  onClick={[Function]}
  tooltipPosition="top"
>
  <GenerateId
    prefix="pf-random-id-"
  >
    <div
      className="pf-c-chip my-chp-cls"
      data-ouia-component-id="OUIA-Generated-Chip-2"
      data-ouia-component-type="PF4/Chip"
      data-ouia-safe={true}
    >
      <span
        className="pf-c-chip__text"
        id="chip_one"
      >
        Chip
      </span>
      <Button
        aria-label="close"
        aria-labelledby="remove_chip_one chip_one"
        id="remove_chip_one"
        onClick={[Function]}
        ouiaId="close"
        variant="plain"
      >
        <button
          aria-disabled={false}
          aria-label="close"
          aria-labelledby="remove_chip_one chip_one"
          className="pf-c-button pf-m-plain"
          data-ouia-component-id="close"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={false}
          id="remove_chip_one"
          onClick={[Function]}
          role={null}
          type="button"
        >
          <TimesIcon
            aria-hidden="true"
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
          >
            <svg
              aria-hidden="true"
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </TimesIcon>
        </button>
      </Button>
    </div>
  </GenerateId>
</Chip>
`;

exports[`Chip closable with tooltip 1`] = `
<Chip
  className="my-chp-cls"
  closeBtnAriaLabel="close"
  component="div"
  id="chip_one"
  isOverflowChip={false}
  isReadOnly={false}
  onClick={[Function]}
  tooltipPosition="top"
>
  <GenerateId
    prefix="pf-random-id-"
  >
    <div
      className="pf-c-chip my-chp-cls"
      data-ouia-component-id="OUIA-Generated-Chip-3"
      data-ouia-component-type="PF4/Chip"
      data-ouia-safe={true}
    >
      <span
        className="pf-c-chip__text"
        id="chip_one"
      >
        1234567890123456789
      </span>
      <Button
        aria-label="close"
        aria-labelledby="remove_chip_one chip_one"
        id="remove_chip_one"
        onClick={[Function]}
        ouiaId="close"
        variant="plain"
      >
        <button
          aria-disabled={false}
          aria-label="close"
          aria-labelledby="remove_chip_one chip_one"
          className="pf-c-button pf-m-plain"
          data-ouia-component-id="close"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={false}
          id="remove_chip_one"
          onClick={[Function]}
          role={null}
          type="button"
        >
          <TimesIcon
            aria-hidden="true"
            color="currentColor"
            noVerticalAlign={false}
            size="sm"
          >
            <svg
              aria-hidden="true"
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </TimesIcon>
        </button>
      </Button>
    </div>
  </GenerateId>
</Chip>
`;

exports[`Chip overflow 1`] = `
<Chip
  className="my-chp-cls"
  closeBtnAriaLabel="close"
  component="div"
  isOverflowChip={true}
  isReadOnly={false}
  onClick={[Function]}
  tooltipPosition="top"
>
  <GenerateId
    prefix="pf-random-id-"
  >
    <div
      className="pf-c-chip pf-m-overflow my-chp-cls"
      data-ouia-component-id="OUIA-Generated-Chip-1"
      data-ouia-component-type="PF4/OverflowChip"
      data-ouia-safe={true}
      onClick={[Function]}
    >
      <span
        className="pf-c-chip__text"
      >
        4 more
      </span>
    </div>
  </GenerateId>
</Chip>
`;

exports[`Chip readonly 1`] = `
<Chip
  className="my-chp-cls"
  closeBtnAriaLabel="close"
  component="div"
  isOverflowChip={false}
  isReadOnly={true}
  onClick={[Function]}
  tooltipPosition="top"
>
  <GenerateId
    prefix="pf-random-id-"
  >
    <div
      className="pf-c-chip my-chp-cls"
      data-ouia-component-id="OUIA-Generated-Chip-4"
      data-ouia-component-type="PF4/Chip"
      data-ouia-safe={true}
    >
      <span
        className="pf-c-chip__text"
        id="pf-random-id-3"
      >
        4 more
      </span>
    </div>
  </GenerateId>
</Chip>
`;
